### PR TITLE
feat: State flow engine (CMC-644)

### PR DESCRIPTION
**Pull request by sabahirfan**
### JIRA link (if applicable) ###
[CMC-644](https://tools.hmcts.net/jira/browse/CMC-644)

### Change description ###
This PR implements the state flow engine.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

**Review by dharmendrak — APPROVED** [review]

**Review by pd-dudkiewicz — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/reform/unspec/stateflow/StateFlowEngine.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+        public String fullName() {
+            return FLOW_NAME + "." + name();
+        }
+    }
+
+    public StateFlow build() {
+        return StateFlowBuilder.<StateFlowEngine.FlowState>flow(FLOW_NAME)
+            .initial(DRAFT)
+            .transitionTo(CLAIM_ISSUED).onlyIf(claimantIssueClaim)
```
Comment (ckalista) [inline src/main/java/uk/gov/hmcts/reform/unspec/stateflow/StateFlowEngine.java:54]: Could we indent to improve readability? I know the IDEA will line everything up but just find it easier to read if transitionsTo's are indented.

File: src/main/java/uk/gov/hmcts/reform/unspec/stateflow/StateFlowEngine.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+import static uk.gov.hmcts.reform.unspec.stateflow.utils.PredicateUtils.claimantRespondToDefence;
+import static uk.gov.hmcts.reform.unspec.stateflow.utils.PredicateUtils.claimantRespondToRequestForExtension;
+import static uk.gov.hmcts.reform.unspec.stateflow.utils.PredicateUtils.defendantAcknowledgeService;
+import static uk.gov.hmcts.reform.unspec.stateflow.utils.PredicateUtils.defendantAskForAnExtension;
+import static uk.gov.hmcts.reform.unspec.stateflow.utils.PredicateUtils.defendantRespondToClaim;
+import static uk.gov.hmcts.reform.unspec.stateflow.utils.PredicateUtils.schedulerStayClaim;
+
+@Component
+@RequiredArgsConstructor
+public class StateFlowEngine {
```
Comment (ckalista) [inline src/main/java/uk/gov/hmcts/reform/unspec/stateflow/StateFlowEngine.java:30]: Could we have this in its own package? The `uk.gov.hmcts.reform.unspec.stateflow` should be for the `framework`. I would create a new package and have in there:
- StateFlowEngine
- FlowState
- FlowPredicate

File: src/main/java/uk/gov/hmcts/reform/unspec/stateflow/utils/PredicateUtils.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+import java.util.function.Predicate;
+
+import static uk.gov.hmcts.reform.unspec.enums.CaseState.STAYED;
+
+public class PredicateUtils {
+
+    public static final Predicate<CaseData> claimantIssueClaim = caseData ->
+        caseData.getClaimIssuedDate() != null
+            && caseData.getLegacyCaseReference() != null;
```
Comment (ckalista) [inline src/main/java/uk/gov/hmcts/reform/unspec/stateflow/utils/PredicateUtils.java:13]: Wonder if we should stick to the bare minimum and only rely on the most simple data as marker. In this scenario would ClaimIssuedDate not be enough?

File: src/test/java/uk/gov/hmcts/reform/unspec/stateflow/StateFlowEngineTest.java [hunk 4]
```diff
@@ -0,0 +1,214 @@
+package uk.gov.hmcts.reform.unspec.stateflow;
```
Comment (ckalista) [inline src/test/java/uk/gov/hmcts/reform/unspec/stateflow/StateFlowEngineTest.java:1]: Super clean! I like it.

**Review by ckalista — APPROVED** [review]
Exactly how I would have imagined this 👍 Just a couple of suggestions.

File: src/main/java/uk/gov/hmcts/reform/unspec/stateflow/StateFlowEngine.java (hunk 2 above)
Comment (sabahirfan) [reply src/main/java/uk/gov/hmcts/reform/unspec/stateflow/StateFlowEngine.java:30]: Agreed. I am extracting them in next PR

File: src/main/java/uk/gov/hmcts/reform/unspec/stateflow/StateFlowEngine.java (hunk 1 above)
Comment (sabahirfan) [reply src/main/java/uk/gov/hmcts/reform/unspec/stateflow/StateFlowEngine.java:54]: done in the next PR.

File: src/main/java/uk/gov/hmcts/reform/unspec/stateflow/utils/PredicateUtils.java (hunk 3 above)
Comment (sabahirfan) [reply src/main/java/uk/gov/hmcts/reform/unspec/stateflow/utils/PredicateUtils.java:13]: Done in the next PR.

**Pull request merged**